azurerm/helpers/azure/app_service.go

```diff
@@ -212,6 +411,346 @@ func SchemaAppServiceSiteConfig() *schema.Schema {
+				siteAuthSettingsProperties.MicrosoftAccountClientID = utils.String(v.(string))
+			}
+
+			if v, ok := microsoftSetting["client_secret"]; ok {
+				siteAuthSettingsProperties.MicrosoftAccountClientSecret = utils.String(v.(string))
+			}
+
+			if v, ok := microsoftSetting["oauth_scopes"]; ok {
+				input := v.([]interface{})
+
+				oauthScopes := make([]string, 0)
+				for _, param := range input {
+					oauthScopes = append(oauthScopes, param.(string))
+				}
+
+				siteAuthSettingsProperties.MicrosoftAccountOAuthScopes = &oauthScopes
+			}
+		}
+	}
+
+	if v, ok := setting["twitter"]; ok {
+		twitterSettings := v.([]interface{})
+
+		for _, setting := range twitterSettings {
+			if setting == nil {
+				continue
+			}
+
+			twitterSetting := setting.(map[string]interface{})
+
+			if v, ok := twitterSetting["consumer_key"]; ok {
+				siteAuthSettingsProperties.TwitterConsumerKey = utils.String(v.(string))
+			}
+
+			if v, ok := twitterSetting["consumer_secret"]; ok {
+				siteAuthSettingsProperties.TwitterConsumerSecret = utils.String(v.(string))
+			}
+		}
+	}
+
+	return siteAuthSettingsProperties
+}
+
+func FlattenAppServiceAuthSettings(input *web.SiteAuthSettingsProperties) []interface{} {
+	results := make([]interface{}, 0)
+	result := make(map[string]interface{})
+
+	if input == nil {
+		log.Printf("[DEBUG] AuthSettings is nil")
+		return results
+	}
+
+	if input.Enabled != nil {
+		result["enabled"] = *input.Enabled
+	}
+
+	additionalLoginParams := make([]string, 0)
+	if s := input.AdditionalLoginParams; s != nil {
+		additionalLoginParams = *s
+	}
+	result["additional_login_params"] = additionalLoginParams
+
+	allowedExternalRedirectUrls := make([]string, 0)
+	if s := input.AllowedExternalRedirectUrls; s != nil {
+		allowedExternalRedirectUrls = *s
+	}
+	result["allowed_external_redirect_urls"] = allowedExternalRedirectUrls
+
+	if input.DefaultProvider != "" {
+		result["default_provider"] = input.DefaultProvider
+	}
+
+	if input.Issuer != nil {
+		result["issuer"] = *input.Issuer
+	}
+
+	if input.RuntimeVersion != nil {
+		result["runtime_version"] = *input.RuntimeVersion
+	}
+
+	if input.TokenRefreshExtensionHours != nil {
+		result["token_refresh_extension_hours"] = *input.TokenRefreshExtensionHours
+	}
+
+	if input.TokenStoreEnabled != nil {
+		result["token_store_enabled"] = *input.TokenStoreEnabled
+	}
+
+	if input.UnauthenticatedClientAction != "" {
+		result["unauthenticated_client_action"] = input.UnauthenticatedClientAction
+	}
+
+	activeDirectorySettings := make([]interface{}, 0)
+	activeDirectorySetting := make(map[string]interface{})
+
+	if input.ClientID != nil {
+		activeDirectorySetting["client_id"] = *input.ClientID
+	}
+
+	if input.ClientSecret != nil {
+		activeDirectorySetting["client_secret"] = *input.ClientSecret
+	}
+
+	if s := input.AllowedAudiences; s != nil {
+		activeDirectorySetting["allowed_audiences"] = *s
+	}
+
+	result["active_directory"] = append(activeDirectorySettings, activeDirectorySetting)
```

Comment:
we should only add an item to the `activeDirectorySettings` list if there's an AAD application configured; for example:

```
activeDirectorySettings := make([]interface{}, 0)

if input.ClientID != nil {
  activeDirectorySetting := make(map[string]interface{})
	
  if input.ClientID != nil {
    activeDirectorySetting["client_id"] = *input.ClientID
  }
	
  if input.ClientSecret != nil {
    activeDirectorySetting["client_secret"] = *input.ClientSecret
  }	
  if s := input.AllowedAudiences; s != nil {
    activeDirectorySetting["allowed_audiences"] = *s
  }
  activeDirectorySettings = append(activeDirectorySettings, activeDirectorySetting)
}
result["active_directory"] = activeDirectorySettings
```

this means that when users don't specify the `active_directory` block it'll have no values when it's flattened